src/USER-PHONON/dynamical_matrix.cpp

```diff
@@ -0,0 +1,491 @@
+#endif
+    } else if (binaryflag) {
+        fp = fopen(filename,"wb");
+    } else {
+        fp = fopen(filename,"w");
+    }
+
+    if (fp == NULL) error->one(FLERR,"Cannot open dump file");
+
+    file_opened = 1;
+}
+
+/* ----------------------------------------------------------------------
+   create dynamical matrix
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::calculateMatrix()
+{
+    int local_idx; // local index
+    int local_jdx; // second local index
+    int nlocal = atom->nlocal;
+    bigint natoms = atom->natoms;
+    int *mask = atom->mask;
+    int *type = atom->type;
+    double imass; // dynamical matrix element
+    double *m = atom->mass;
+    double **f = atom->f;
+
+    //initialize dynmat to all zeros
+    for (int i=0; i < dynlen; i++)
+        for (int j=0; j < dynlen; j++)
+            dynmat[i][j] = 0.;
+
+    energy_force(0);
+
+    if (comm->me == 0 && screen) fprintf(screen,"Calculating Dynamical Matrix...\n");
+
+    for (int i=1; i<=natoms; i++){
+        local_idx = atom->map(i);
+        //if (screen) fprintf(screen, "local idx = %i on proc %i with %i local\n",local_idx, comm->me, nlocal);
+        if (local_idx >= 0){
+            for (int alpha=0; alpha<3; alpha++){
+                displace_atom(local_idx, alpha, 1);
+                energy_force(0);
+                for (int j=1; j<=natoms; j++){
+                    local_jdx = atom->map(j);
+                    if (local_jdx >= 0  && local_idx < nlocal){
+                        for (int beta=0; beta<3; beta++){
+                            dynmat[(i-1)*3+alpha][(j-1)*3+beta] += -f[j-1][beta];
+                        }
+                    }
+                }
+                displace_atom(local_idx,alpha,-2);
+                energy_force(0);
+                for (int j=1; j<=natoms; j++){
+                    local_jdx = atom->map(j);
+                    if (local_jdx >= 0 && local_idx < nlocal){
+                        for (int beta=0; beta<3; beta++){
+                            if (atom->rmass_flag == 1)
+                                imass = sqrt(m[i] * m[j]);
+                            else
+                                imass = sqrt(m[type[i]] * m[type[j]]);
+                            //dynmat has length dynlen
+                            //dynlen = (group->count(igroup))*3;
+                            //currently dynmat is being called to natoms*3
+                            //Also with this implementation
+                            //I am not recovering the correct dynamical matrix
+                            //if I have more than 1 core
+                            dynmat[(i-1)*3+alpha][(j-1)*3+beta] -= -f[j-1][beta];
+                            dynmat[(i-1)*3+alpha][(j-1)*3+beta] /= (2 * del * imass);
+                            dynmat[(i-1)*3+alpha][(j-1)*3+beta] *= conversion;
+                        }
+                    }
+                }
```

Comment:
if dynmat changed to be a (local) 3 x dynlen matrix, then it must be reduced here, and then added to the global dynamical matrix. This would reduce the memory consumption for large systems significantly.

Reply:
yes. you can reduce memory use even more by outputting the dynamical matrix here directly. in LAMMPS we try to distribute data as much as possible, and thus we try to avoid allocations of size `atom->natoms` (but that is not always possible). `atom->natoms` with a constant factor is similar, but `atom->natoms * atom->natoms` is a much bigger problems, as memory use now grows quadratic with the number of atoms and cannot be reduced by parallelization.
...and if quadratic size cannot be avoided, then at least it should be only on one MPI rank (i.e. use `MPI_Reduce()` instead of `MPI_Allreduce()`), i.e. the one writing the file.

The whole matrix would really only be needed, e.g. if you include some postprocessing, and even then with additional effort, that could be distributed.

But then again, any size of a system, where this would matter, is probably taking such a long time to compute, that nobody would want to do it in the first place. Hence my suggestion, to run the force computation once to record the time it take for one force compute and then write out an estimate for the total time it would take to compute the matrix.

Reply:
I should definitely do MPI_Reduce instead of all reduce. I wrote that back when I was learning MPI...

    But then again, any size of a system, where this would matter, is probably taking such a long time to compute, that nobody would want to do it in the first place. Hence my suggestion, to run the force computation once to record the time it take for one force compute and then write out an estimate for the total time it would take to compute the matrix.

Should I make a check_comp_time method?

Reply:
That seems a bit overkill to me. i would just leave the one call to the force computation in the `setup()` method (that seems the best place for it) and measure its time. 

if you then use the following code (adapted from `src/run.cpp`:
```
   timer->init();
    timer->barrier_start();
    calculateMatrix();
    timer->barrier_stop();
```
If you keep the timer calls in the force routine, as they were in verlet.cpp or similar, then you have a nice breakdown of how the time is spent on the different force computations, in the summary output from the Finish class.

The call in the setup function could in the future also be used to get per MPI rank timer information to be used for load balancing, although that is something that can be done outside the command using a run 0 and a balance command. There are lots of additional 'gimmicks' that can be added later, a timeout (for when running with batch systems), a restart feature (for very large systems) etc.